src/libopensc/card-piv.c

```diff
@@ -467,205 +791,1921 @@ static struct sc_card_driver piv_drv = {
 	NULL, 0, NULL
 };
 
+static int piv_get_cached_data(sc_card_t * card, int enumtag, u8 **buf, size_t *buf_len);
+static int piv_cache_internal_data(sc_card_t *card, int enumtag);
+static int piv_logout(sc_card_t *card);
 static int piv_match_card_continued(sc_card_t *card);
 
-static int
-piv_find_obj_by_containerid(sc_card_t *card, const u8 * str)
+/* compare  sc_asn1_read_tag to expected tag_long */
+/*  after #2079  can be converted to  inline  "if ((cla<<24 ! tag) == tag_long))" */
+static int piv_is_expected_tag(unsigned int cla, unsigned int tag, unsigned int tag_long)
 {
-	int i;
+	unsigned int tag_read = tag;
 
-	LOG_FUNC_CALLED(card->ctx);
-	sc_log(card->ctx, "str=0x%02X%02X\n", str[0], str[1]);
+	/* UNIVERSAL is 0,  PRIVATE is both APPLICATION and CONTEXT */
+	tag_read |= ((cla & SC_ASN1_TAG_APPLICATION) ? SC_ASN1_APP : 0);
+	tag_read |= ((cla & SC_ASN1_TAG_CONTEXT) ? SC_ASN1_CTX : 0);
+	tag_read |= ((cla & SC_ASN1_TAG_CONSTRUCTED) ? SC_ASN1_CONS : 0);
+	return  (tag_read == tag_long);
+}
 
-	for (i = 0; piv_objects[i].enumtag < PIV_OBJ_LAST_ENUM; i++) {
-		if ( str[0] == piv_objects[i].containerid[0] && str[1] == piv_objects[i].containerid[1])
-			LOG_FUNC_RETURN(card->ctx, i);
+#ifdef ENABLE_PIV_SM
+static void piv_inc(u8 *counter, size_t size)
+{
+	unsigned int c = 1;
+	unsigned int b;
+	int i;
+	for (i = size - 1; c != 0 && i >= 0; i--){
+			b = c + counter[i];
+			counter[i] = b & 0xff;
+			c = b>>8;
 	}
```

Comment:
Can you reuse existing `sm_incr_ssc`? I think it does exactly the same, regardless the confusing name.

Reply:
> Can you reuse existing sm_incr_ssc? I think it does exactly the same, regardless the confusing name.
Maybe. But @frankmorgner wants a major overhaul of the PIV SM code, which would include this. So for now I would like to keep the working version of the PIV SM code, so all the other changes in #2053  are committed before addressing Frank's SM request.

Reply:
`sm_incr_ssc` means "increment send sequence counter". the sequence counter is incremented for each message in either direction. The GP/EAC model uses one counter that is incremented for each command and for each response. The PIV model additionally changes the first byte to `80` in case of a response (@dengert implemented this as two seperate counters). However, incrementation is identical.

`sm_incr_ssc()` is included in libsm, which is currently only included in the default SM module (`src/smm`), but not in libopensc...